pkg/dataset/dataset.go

```diff
@@ -105,6 +108,98 @@ func LoadCSV(r io.Reader) (*mat64.Dense, error) {
 	return mat64.NewDense(rows, cols, mxData), nil
 }
 
+// LoadLRN reads data from a .lrn file.
+// See the specification here: http://databionic-esom.sourceforge.net/user.html#Data_files____lrn_
+func LoadLRN(reader io.Reader) (*mat64.Dense, error) {
+	const DATA_COL = 1
+	const (
+		HEADER_SIZE = iota
+		HEADER_COLS
+		HEADER_TYPES
+		HEADER_NAMES
+		HEADER_ROWS
+	)
+
+	var rows, cols int
+	var mxData []float64
+	headerRow := 0
+	columnTypes := []int{}
+	valueRow := 0
+
+	scanner := bufio.NewScanner(reader)
+	for scanner.Scan() {
+		line := scanner.Text()
+		if strings.HasPrefix(line, "#") { // comment
+			continue
+		} else if strings.HasPrefix(line, "%") { // header
+			headerLine := strings.TrimPrefix(line, "% ")
+			if headerRow == HEADER_SIZE { // rows
+				rows64, err := strconv.ParseInt(headerLine, 10, 64)
+				if err != nil {
+					fmt.Println(err)
+					return nil, fmt.Errorf("Dataset size information missing")
+				}
+				rows = int(rows64)
+			} else if headerRow == HEADER_COLS { // cols
+				// discard
+			} else if headerRow == HEADER_TYPES { // col types
+				colTypes := strings.Split(headerLine, "\t")
+				for _, colType := range colTypes {
+					// this seems to happen in real .lrn files
+					if len(colType) == 0 {
+						continue
+					}
+					ct, err := strconv.ParseInt(colType, 10, 64)
+					if err != nil {
+						return nil, err
+					}
+					columnTypes = append(columnTypes, int(ct))
+					// we're interested in data columns only
+					if ct == DATA_COL {
+						cols++
+					}
+				}
+				// allocate data matrix because we know rows and cols now
+				mxData = make([]float64, rows*cols)
+			} else if headerRow == HEADER_NAMES { // col names
```

Comment:
Well, strictly speaking it is not needed, but I think it's more readable if we keep it.